### ISSUE TITLE
Hasher should be re-usable via a reset() method

### ISSUE THREAD

**Issue by harningt** [google/guava#2396]
One use-case for a Hasher is implementing PBKDF2... specifically using the new hmac set of hashes.

The PBKDF2 algorithm calls for performing a hash using the same MAC a very large number of times. Directly with the Java Mac interface object, doFinal resets the state so that the Mac can be re-used... resulting in only the byte array being regenerated each time (even doFinal with output byte array in JDK allocates due to the SPI not having an 'output' engineFinalize call).

Making a 'reset' method for many of the Hasher implementations seems to be straightforward in behavior:

- State reset in the object however required (for Mac, there exists a 'reset' call which would be called if doFinal wasn't recently called)
- Future calls to put more data in will succeed
- Future call to hash() will succeed

Refining what repeated calls to hash() means would also be useful - right now it is declared as 'undefined behavior'. It could be declared that hash() performs 'reset' implicitly (doFinal for Mac does this).

Implementing this reset behavior would result in only 1 extra allocation per iteration:
- The wrapper HashCode around the byte array returned by doFinal()

Another nice enhancement would be to have a function call to perform a hash and directly stream its bytes to a byte[] or ByteBuffer, avoiding the intermediate object.

Yet another wishlist item would be PBKDF2 directly implemented - but it seems pretty specialized and requires many inputs to make sense... not quite as straightforward as hmac design-wise.

**Comment by damadden** [comment]
I ran into the issue outlined recently.  To perform any hash, a new Hasher object needs to be created.  Doing this very many times (millions+) results in an inordinate amount of garbage collection.  In my implementation, the execution lasted 24 minutes and was dominated by almost 18 minutes of GC alone.  When I made the change to simply clear() the backing ByteBuffer and reset the relevant variables in the hash function (I was using the Murmur3_128HashFunction), I could use a single Hasher instance and GC was virtually non-existent.

**Comment by kluever** [comment]
Thanks for the motivating use case. Adding a method to an interface (`Hasher`) is a breaking change pre-Java8, so we'd have to be _extra_ motivated to do so.

So when you made the change to clear the buffers, are you saying your execution of the same chunk of code when from 24 minutes to 6 minutes?!? What type of environment is this happening on?

**Comment by ronshapiro** [comment]
We _could_ add a method on `AbstractStreamingHasher` without making a breaking change, but that would be a strange place to put a method for an API.

`Hasher` is `@Beta` though, so adding a method and then implementing it in `AbstractHasher` to be a noop isn't the worst thing in the world.

**Comment by kluever** [comment]
-1 on adding it to the implementation helper classes. It really belongs on `Hasher` directly if we're going to add it.

`HashFunction` and `Hasher` are indeed `@Beta` (added in 11.0), and we did make breaking changes to them in Guava 15.0, but I'd still feel really bad about making a breaking change now.

The OP could just define their own `reset()` method on their Hasher, and just maintain references to their specific types. The 2nd poster is in a more tricky situation, and if performance is a huge concern, then maybe it's worth forking our implementation? I'm not really sure...

**Comment by lowasser** [comment]
Also note that "murmur3_128 is causing too much GC" is a separate issue that can be at least partially addressed without changing API surface, by redesigning it to allocate *less* even if we can't drive that down to zero.

**Comment by damadden** [comment]
Yes - execution time went from 24 mins to 6 mins when the buffers were
cleared instead of new ones created.  This practically eliminated the GC,
which was proving to be excessive otherwise.  Rough calculations showed
that hashing the data (~300 GiB) using a new buffer for each value hashed
resulted in the generation of ~1.78 TiB of data - 35 *(columns)* x 342
Million *(rows)* x 160 B *(buffer size)*.

*System details:*
1 of 4 data nodes in a Hadoop cluster
  - RAM: 128 GiB total system, JVM configured for 12.8 GiB Xmx
  - CPU: 20 physical CPUs (40 cores with hyper-threading), JVM
limited/bound to 20 cores
  - Processor type: Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz
  - Java: OpenJDK 8

*Workload (performed by 1 node):*
  - Hashing performed on all the values in a table with 35 columns and over
342 Million rows
  - Values are of types INT, BIGINT and DOUBLE only
  - Approximately half of the heap is occupied by large, persistent objects
(hard references)

**Comment by damadden** [comment]
Since version 21, Guava has required Java 8, which has support for `default` in interface methods.  Surely we can implement this change now without breaking legacy?

**Comment by jrtom** [comment]
Guava doesn't overall require Java 8; it has two "flavors", one which requires 8, the other of which requires 7: https://github.com/google/guava/#guava-google-core-libraries-for-java
